extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+    }
+    buf.push(normalizeChar(c));
+  }
+  return buf.join('');
+};
+
+/**
+ * findSentences find sentences from node and returns a list of TextRange.
+ * @param {!Node} node
+ * @param {!Array<string>} sentences
+ * @return {?Array<!TextRange>}
+ */
+export const findSentences = function(node, sentences) {
+  const scanner = new TextScanner(node);
+  const matches = [];
+  for (let senIdx = 0; senIdx < sentences.length; senIdx++) {
+    const sen = normalizeString(sentences[senIdx]);
+    if (!sen) {
+      continue;
+    }
+    // BM-algorithm with bad-character rules.
+    const skipTable = {};
+    for (let i = 0; i < sen.length; i++) {
+      const c = sen[i];
+      skipTable[c] = sen.length - 1 - i;
+    }
+    const buf = new CircularBuffer(sen.length);
+    let index = -1;
+    let nextIndex = sen.length - 1;
+    while (true) {
+      const pos = scanner.next();
+      if (pos == null) {
+        // mismatch
+        return null;
+      }
+      if (pos == posDomDelimiter || skipCharRe.test(pos.char)) {
+        continue;
+      }
+      buf.add(pos);
+      index++;
+      if (index < nextIndex) {
+        continue;
+      }
+      let ok = true;
+      for (let j = 0; j < sen.length; j++) {
+        const c = normalizeChar(buf.get(sen.length - j - 1).char);
+        if (sen[sen.length - 1 - j] == c) {
+          continue;
+        }
+        ok = false;
+        let skip = skipTable[c];
+        if (skip == null) {
+          skip = sen.length;
+        }
+        skip -= j;
+        if (skip < 1) {
+          skip = 1;
+        }
+        nextIndex += skip;
+        break;
+      }
+      if (ok) {
+        const endPos = buf.get(sen.length - 1);
+        matches.push(new TextRange(
+            buf.get(0), new TextPos(endPos.node, endPos.offset + 1)));
+        break;
+      }
+    }
+  }
+  return matches.length > 0 ? matches : null;
+};
+
+/**
+ * @param {!Array<!TextRange>} ranges
+ * @return {!Array<!Element>} A list of marked nodes.
+ */
+export const markTextRangeList = function(ranges) {
+  ranges = concatContinuousRanges(ranges);
+  const marked = [];
+  for (let i = 0; i < ranges.length; i++) {
+    const r = ranges[i];
+    markTextRange(r.start, r.end, ranges, i, marked);
+  }
+  return marked;
+};
+
+/**
+ * @param {!Array<!TextRange>} ranges
+ * @return {!Array<!TextRange>}
+ */
+const concatContinuousRanges = function(ranges) {
+  const ret = [];
+  let prev = null;
+  for (let i = 0; i < ranges.length; i++) {
+    const r = ranges[i];
+    if (prev && prev.end.node == r.start.node &&
+        prev.end.offset == r.start.offset) {
+      prev.end = r.end;
+      continue;
+    }
+    prev = r;
+    ret.push(r);
+    continue;
+  }
+  return ret;
+};
+
+/**
+ * @param {!TextPos} start
+ * @param {!TextPos} end
+ * @param {!Array<TextRange>} ranges Other ranges
+ * @param {number} idx
+ * @param {!Array<!Element>} marked
+ */
+const markTextRange = function(start, end, ranges, idx, marked) {
+  while (true) {
```

Comment:
nit: There's several exit points (`return`, `break`) here. Consider refactoring so that exit points are grouped together, if possible.